### PR TITLE
[gui] Move resolving of fonts/colors into the skin object

### PR DESCRIPTION
**Pull request by phil65**
This basically re-applies changes from https://github.com/xbmc/xbmc/pull/4743 and makes it work / compile with today´s codebase. To see what was done perhaps best to look at the linked PR, since this here is squashed down to one single commit. Can try to make it bisectable in case this PR gets considered.
Hopefully brings colors/fonts support for add-ons a step closer.
@jmarshallnz 	, sorry to bug you, could you sum up what still needs to be done here?
From quick testing I couldnt spot any issues.
No idea who to ping... perhaps @Montellese , or @FernetMenta 	for architecture.

### CODE REVIEW

**Comment by phil65** [comment]
Resolved conflicts.
Since noone reviewed yet, perhaps @notspiff could give an opinion?

File: xbmc/addons/Skin.cpp [hunk 1]
```diff
@@ -27,6 +27,11 @@
 #include "filesystem/Directory.h"
 #include "filesystem/File.h"
 #include "filesystem/SpecialProtocol.h"
+
+#include "guilib/GUIControlFactory.h" // for colors
```
Comment (notspiff) [inline xbmc/addons/Skin.cpp]: factory include seems rather misplaced. include the headers directly imho.

File: xbmc/guilib/GUIFontManager.h [hunk 2]
```diff
@@ -62,12 +62,11 @@ class GUIFontManager : public IMsgTargetCallback
 public:
   GUIFontManager(void);
   virtual ~GUIFontManager(void);
-
+
```
Comment (notspiff) [inline xbmc/guilib/GUIFontManager.h]: no

**Review by notspiff — COMMENTED** [review]
In general I personally never use pointers (and certainly not shared pointers!) unless i need the optionality but this is (mostly) a style thing that i guess is used around here.

A nice cleanup all-around. I haven't runtime tested or such but i suspect you have ;)

File: xbmc/addons/Skin.cpp (hunk 1 above)
Comment (phil65) [inline xbmc/addons/Skin.cpp]: Agreed. Not even sure if this include is needed at all, seems to build fine without it (at least on linux)
Might be some relict from the old PR.

**Comment by phil65** [comment]
jenkins build this please.

**Comment by Paxxi** [comment]
I don't like all of that shared_ptr passing but I've only looked at it on github. I'll try and have a proper look at it tomorrow or Thursday and see.

**Comment by phil65** [comment]
@Paxxi ping

**Comment by Rechi** [comment]
@Paxxi ping
@phil65 this needs a rebase

File: xbmc/guilib/GUIControlFactory.cpp [hunk 3]
```diff
@@ -126,7 +125,7 @@ std::string CGUIControlFactory::TranslateControlType(CGUIControl::GUICONTROLTYPE
   return "";
 }
 
-CGUIControlFactory::CGUIControlFactory(void) = default;
+CGUIControlFactory::CGUIControlFactory(GUIResourceProviderPtr provider) = default;
```
Comment (Rechi) [inline xbmc/guilib/GUIControlFactory.cpp]: error: `CGUIControlFactory::CGUIControlFactory(GUIResourceProviderPtr)` cannot be defaulted

File: xbmc/addons/Skin.h [hunk 4]
```diff
@@ -183,6 +185,30 @@ class CSkinInfo : public CAddon
...
+  /*! \brief load Fonts.xml for the skin, adding fonts to the font manager
+ +   \param fontSet the name of the <fontset> to load.
+ +   */
+  void LoadFonts(const std::string &fontSet);
+
+  /*! \brief fetch a font from the skin
+ +   \param fontName the name of the font (from Fonts.xml) to fetch
+ +   \return a CGUIFont pointer with the font, NULL if not found
+ +   */
+  virtual CGUIFont *GetFont(const std::string &fontName) const;
```
Comment (Rechi) [inline xbmc/addons/Skin.h]: `override` modifier missing

File: xbmc/addons/Skin.h [hunk 5]
```diff
@@ -183,6 +185,30 @@ class CSkinInfo : public CAddon
...
+ +   Loads the default system colors, default skin colors, then chosen skin colors.
+ +   \param colorFile color file to load
+ +   */
+  void LoadColors(const std::string &colorFile);
+
+  /*! \brief retrieve a color by name.
+ +   \param color the name of the color to retrieve
+ +   \return the color value, if found, else 0xffffffff (white).
+ +   */
+  virtual color_t GetColor(const std::string &color) const;
```
Comment (Rechi) [inline xbmc/addons/Skin.h]: `override` modifier missing

**Comment by Paxxi** [comment]
I don't really remember my objection but I'm sure this is fine

**Comment by phil65** [comment]
865th rebase done.

**Comment by notspiff** [comment]
for the love of everything warm and fuzzy merge when it passes!

**Comment by jmarshallnz** [comment]
Hi! By the looks labelInfoMono.font needs to be dealt with in the same way labelInfo.font is further up.

**Comment by phil65** [comment]
> Hi! By the looks labelInfoMono.font needs to be dealt with in the same way labelInfo.font is further up.

Rebased, hopefully all fine now.

**Comment by phunkyfish** [comment]
PR closed as it appears to be abandoned by the author. 

If the author would like it kept open please just request so on the PR. If you feel there is still merit in merging this PR you can request the label be changed to `PR Cleanup: Backburner`.

**Pull request closed without merging**